libs/constrained-generators/src/Constrained/GenT.hs

```diff
@@ -332,3 +332,24 @@ frequency2 (n, g1) (m, g2)
         Right x -> pure x
 
 -- ======================================
+
+-- | Temporarily extend the stack while executing 'm', and revert to the old stack if successful
+push :: forall m a. MonadGenError m => [String] -> GenT GE a -> GenT m a
+push [] m = dropGen m
+push (x : xs) m =
+  case explain (x NE.:| xs) m of
+    GenT f -> (GenT g)
+      where
+        g :: GenMode -> Gen (m a)
+        g mode = do
+          result <- f mode
+          case result of
+            Result (_ : ys) a -> pure $ runGE (Result ys a)
+            other -> pure $ runGE other
+
+pushGE :: forall a. [String] -> GE a -> GE a
+pushGE [] x = x
+pushGE (x : xs) m = do
+  case explain (x NE.:| xs) m of
+    Result (_ : ys) a -> Result ys a
+    other -> other
```

Comment:
I don't get at all why this is necessary. This is what `explain` already does??

Reply:
explain, adds an explanation, and it is never removed.  This code adds a explanation for one computation, and then if it succeeds, it is rolled back to the state before.

Reply:
Right, that's a bug in `explain`. It shouldn't do that for `Result`!